### ISSUE TITLE
Can't call macros_info() from .md file mkdocs: Error reading page 'info.md': 'macros_info' is undefined

### ISSUE THREAD

**Issue by MikeLemo1** [fralau/mkdocs-macros-test#1]
Hi I've installed and defined:
```yml
 - macros:
      modules: [mkdocs-macros-test:mkdocs_macros_test]
```

in mkdocs.yml file and when I add the line: `{{ macros_info() }}`  to an empty info.md file I get the following error and mkdocs won't serve the doc:

> ERROR    -  Error reading page 'info.md': 'macros_info' is undefined

Any Idea why does that happen?

Here's the mkdocs.yml file content:
```yml

site_name: My Docs
theme: 
  name: material
  # ---Might be exclusive to Material theme---
  features:
    - search.suggest
    - search.highlight
    - content.tabs.link
    
    - content.code.copy # Enables  "CODE_BLOCK" Copying
    - content.code.select # Enables "CODE_BLOCK" selection 

    - content.action.edit # ?
    - content.action.view # ?

    # - navigation.tabs
    # - navigation.tabs.sticky
    - navigation.expand  
    - navigation.path 

    - toc.integrate # Expands web space over TOC thingy

  language: en

  # ---Might be exclusive to Material theme---
  palette:
    - scheme: slate 
      # toggle:
      #   icon: material/toggle-switch
      #   name: Switch to light mode    
      primary: teal
      accent: purple    

    - scheme: default
      toggle:
        icon: material/toggle-switch-off-outline 
        name: Switch to dark mode
      primary: teal
      accent: lime 
# readthedocs


nav:
  - Home: index.md
  - Initial installation: SoftwareDoc.md
  - Tutorial:
    - VSCode_doc.md
  - MkDocs Doc: MkDocs_docs.md
  - Notebook page: jupyer_sandbox.ipynb
  

  - Programming: Programming_Doc.md

  - info.md
  # - Python file: python_script.py




plugins:

  - search
  - table-reader
  - external-markdown
  - autorefs
  - macros:
      modules: [mkdocs-macros-test:mkdocs_macros_test]
      # modules: 
      #  - my_definitions
      # module_name: definitions
      # include_dir: include
      # on_error_fail: true
      # on_undefined: strict

  # For image editing:
  - glightbox
  - mkdocs-jupyter:
      theme: dark
      ignore: ["./jupyer_sandbox.ipynb", "./Jupiter_journey.ipynb", ACDC_Experinments.ipynb]
      include_requirejs: true


extra_javascript:
#  Enables mathjax latex
  - https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-AMS-MML_HTMLorMML


# ---Might be exclusive to Material theme---
markdown_extensions:

  # - external-markdown # Might break some stuff

  - footnotes
  # - toc(permalink=true)
  - toc:
      permalink: "##"

  - mdx_math
  - pymdownx.snippets
  - pymdownx.smartsymbols

  # Embed mathjax equation notations
  - pymdownx.arithmatex

  # sub snd super scripting enabled 
  - pymdownx.caret
  - pymdownx.mark
  - pymdownx.tilde

  # Extension to show keyboard shourtcut:
  - pymdownx.keys

  # Used for codeblock design 
  - pymdownx.highlight:
      # Enables line numbering option
      linenums: true
      # Wat?
      anchor_linenums: true
      # Automatically adds a title to codeblock based of setup
      auto_title: true


  - pymdownx.superfences

  # Integrate pdf files into MD
  - pymdownx.pathconverter:
      base_path: './' # default: ''
      relative_path: '' # default ''
      absolute: true # default: false
      tags: 'a script img link object embed'


```

**Comment by fralau** [comment]
Thanks! What I would do, would be to make a copy of your website, and remove all other calls to plugins, extensions, etc., to better isolate the problem.

Just to make sure, did you install [mkdocs-macros-plugin](https://github.com/fralau/mkdocs_macros_plugin)?

Also, is there a reason why you are you calling `mkdocs_macros_test` ?

**Comment by MikeLemo1** [comment]
Hi I just did what you suggested and separated and tested the function of the other plugins and turns out that the extension " external markdown" caused macro plugin to malfunction  for some reason and when building mkdocs in the cmd it shows me this if it says something: 

> WARNING  -  [09:56:28] "GET /assets/javascripts/bundle.220ee61c.min.js HTTP/1.1" code 404

yes [mkdocs-macros-plugin](https://github.com/fralau/mkdocs_macros_plugin) was installed and calling mkdocs_macros_test because I wanted to test the macros plugin to findout why it doesn't work. 

It's fine for now that the "External markdown" plugin doesn't work because it still can't add referenced sections from local docs but expecting it to have this ability in the future which will make it quite useful, so I might be interested in fixing it in the near future if I'll have enough time, you think there is any idea and quick fix for that problem?

**Comment by fralau** [comment]
The problem is quite explainable: **[external markdown uses directives that also rely on the jinja2 templating system](https://pypi.org/project/mkdocs-embed-external-markdown/), the same language that mkdocs-macros uses.** It is that plugin that fails (not mkdocs-macros), because it can't interpret any macros that are not its own.

Perhaps it could work, if that plugin did not fail when finding macros it cannot interpret? 

On the other hand, the whole idea of mkdocs-macros is to implement what this extension does with a macro (python function). 

Indeed, it is provided out of the box with the standard [`{% include %}`](https://jinja.palletsprojects.com/en/3.1.x/templates/#include) directive provided by Jinja2. `include` interprets templates, but that should be an advantage, not be a problem, at least most of the time.

Or else you can write your own function very easily.

Here is the code of the function directly from that [extension's code](https://github.com/fire1ce/pypi-publishing-actions/blob/main/external_markdown/plugin.py):

```python
def external_markdown(self, url, section_name):
    self.is_valid_url(url)
    if section_name:
        markdown = self.get_markdown_from_url(url)
        if markdown:
            return self.get_section_from_markdown(markdown, section_name, url)
    else:
        return self.get_markdown_from_url(url)
```

It should be easy to implement, either as a [macro module](https://mkdocs-macros-plugin.readthedocs.io/en/latest/macros/), or a [pluglet](https://mkdocs-macros-plugin.readthedocs.io/en/latest/pluglets/).

In other words, it is a strategic choice: if you use mkdocs-macros, the external-markdown extension (or any plugin or extension that uses jinja2) becomes redundant; since all those plugins and extensions are particular implementations of what mkdocs-macros was created to do, in general.

**Comment by MikeLemo1** [comment]
Huh ok Now I somewhat understand why "external-markdown" causes "mkdocs-macros" to fail because of the jinja2 interpretation clash guess I'll leave "external-markdown" not that you gave a decent idea, explanation and close reference to the solution which I'll try to completely workon and probably touch it in the next 2 week since I don't have much time with all the work.
And let me tell you something the name if the extension mkdocs-macros gives me somewhat of hard time understanding it's purpose, intended use and the general idea of how and for what it is intended to be used,
only after specific reading about it for the last two days gives you a clearer idea to what it's suppose to do 
because be honest with you it's weird that you have to add this extension to mkdocs instead of having this functionality builtin... maybe the official mkdocs team should add this extension to the app by default also in term of the documentation.

**Comment by fralau** [comment]
@MikeLemo1 I understand where you are coming from, and why it can be frustrating. As you perceived correctly the plugins that use jinja2 functions are likely to be incompatible with each other. That's in the nature of jinaj2.

I also believe that the creators of mkdocs did a good job, and I wouldn't blame them for not including what mkdocs-macros is doing -- because it was a novelty. With the benefit of hindsight, it might seem obvious, but it wasn't at first. If I did it, it is because no one had had had this idea before; or if someone had, they had not acted on it. 

One of the reasons that led me to mkdocs-macros (a general purpose implementation of jinja2 templating) is that I looked at the plugins that used jinja2 and I noticed that all these distinct plugins could be generalized into one! And that would do away with the incompatibilities. If all these plugins were re-implemented as mkdocs-macros "pluglets", that could solve the problem 🤔 .

**Comment by MikeLemo1** [comment]
Luckily you recruited your self to step in and make after you open the gates of Python to mkdocs your alcohol consumption ability is your limit. 
A few things to refine to be able to add this officially but not far at all
